### PR TITLE
Prevent scheduler settings to be removed

### PR DESCRIPTION
**Pull request by imobachgs**
* Fixes [bsc#968192](https://bugzilla.suse.com/show_bug.cgi?id=968192).
* SystemSettings refactoring.
* Added coveralls.

### CODE REVIEW

File: src/modules/SystemSettings.rb [hunk 1]
```diff
@@ -15,210 +15,304 @@ module Yast
...
-      end
+    KERNEL_SYSRQ_FILE = "/proc/sys/kernel/sysrq".freeze
+
+    # Return kernel configuration value for SysRq keys
+    #
+    # The value is read from /proc/sys/kernel/sysrq
+    #
+    # @return [String] Configuration value; returns an empty string if it's not set.
+    def kernel_sysrq
+      return @kernel_sysrq unless @kernel_sysrq.nil?
```
Comment (jreidinger) [inline src/modules/SystemSettings.rb]: kernel_sysrq is string, so simple `return @kernel_sysrq if @kernel_sysrq` is enough

File: src/modules/SystemSettings.rb [hunk 2]
```diff
@@ -15,210 +15,304 @@ module Yast
...
+      content = SCR.Read(path(".target.string"), KERNEL_SYSRQ_FILE).to_s
+      @kernel_sysrq ||= content.split("\n")[0] || ""
```
Comment (jreidinger) [inline src/modules/SystemSettings.rb]: useless `||=` as above you are sure that it is nil, so simple `=` can be used.

File: src/modules/SystemSettings.rb [hunk 3]
```diff
@@ -15,210 +15,304 @@ module Yast
...
+    end
 
-      true
+    # Return sysctl configuration value for SysRq keys
+    #
+    # @return [String] Configuration value; returns an empty string if it's not set.
+    def sysctl_sysrq
+      return @sysctl_sysrq unless @sysctl_sysrq.nil?
```
Comment (jreidinger) [inline src/modules/SystemSettings.rb]: same here with unnecessary nil?

File: src/modules/SystemSettings.rb [hunk 4]
```diff
@@ -15,210 +15,304 @@ module Yast
...
-    # Kernel param 'elevator'
-    def GetIOScheduler
-      @elevator
+    # Activate IO scheduler
+    #
+    # @see activate_scheduler
+    def activate_scheduler
+      return if GetIOScheduler().nil?
+
+      Yast.import "Bootloader"
```
Comment (jreidinger) [inline src/modules/SystemSettings.rb]: above it uses Bootloader without import. So global import is needed or local import at each place?

File: src/modules/SystemSettings.rb [hunk 5]
```diff
@@ -15,210 +15,304 @@ module Yast
...
+
+      log.info("Activating scheduler: #{new_elevator}")
+      # set the scheduler
+      Bootloader.modify_kernel_params("elevator" => new_elevator)
+      # set bootloader configuration as 'changed' (bsc#968192)
+      Bootloader.proposed_cfg_changed = true
+
+      # activate the scheduler for all disk devices
+      return if new_elevator == :missing
+      Dir["/sys/block/*/queue/scheduler"].each do |f|
```
Comment (jreidinger) [inline src/modules/SystemSettings.rb]: does it make sense to do it in installation in insts-sys?

File: src/modules/SystemSettings.rb [hunk 6]
```diff
@@ -15,210 +15,304 @@ module Yast
...
+      return if GetIOScheduler().nil?
+
+      Yast.import "Bootloader"
+      new_elevator = GetIOScheduler() == "" ? :missing : GetIOScheduler()
+
+      log.info("Activating scheduler: #{new_elevator}")
+      # set the scheduler
+      Bootloader.modify_kernel_params("elevator" => new_elevator)
+      # set bootloader configuration as 'changed' (bsc#968192)
+      Bootloader.proposed_cfg_changed = true
```
Comment (jreidinger) [inline src/modules/SystemSettings.rb]: if it is run standolone on runing system, then it have to write Bootloader configuration ( but only there to not slow down installation ).

File: src/modules/SystemSettings.rb [hunk 7]
```diff
@@ -15,210 +15,304 @@ module Yast
...
+    # bootloader configuration is written at the end of the first stage.
+    #
+    # @see Bootloader#Write
+    # @see Write
+    def write_scheduler
+      return true unless Mode.normal
+      Yast.import "Bootloader"
+      Bootloader.Write
+      true
+    end
```
Comment (jreidinger) [inline src/modules/SystemSettings.rb]: in general whole this class looks to me like it deserve two objects: Scheduler and Sysrq one, with its read/write/activating/etc.

File: src/modules/SystemSettings.rb (hunk 6 above)
Comment (jreidinger) [inline src/modules/SystemSettings.rb]: ah, write is below. To be honest this is first module where I see this activate/write split. not much intuitive.

File: src/modules/SystemSettings.rb (hunk 7 above)
Comment (imobachgs) [inline src/modules/SystemSettings.rb]: Yes, I agree, but I wanted to limit the scope of the refactoring :)

File: src/modules/SystemSettings.rb (hunk 6 above)
Comment (imobachgs) [inline src/modules/SystemSettings.rb]: It's done in the Write method.

File: src/modules/SystemSettings.rb (hunk 5 above)
Comment (imobachgs) [inline src/modules/SystemSettings.rb]: AFAIK, the old code did, so I kept it. I guess that we could take advantage of the new scheduler.

File: src/modules/SystemSettings.rb (hunk 4 above)
Comment (imobachgs) [inline src/modules/SystemSettings.rb]: Good point. I would move the import to the `#main` method.

Comment (jreidinger) [inline src/modules/SystemSettings.rb]: if you move it to main, then you can move it to top level, as main is called as part of file initialization, so require will call it

File: src/modules/SystemSettings.rb [hunk 8]
```diff
@@ -262,7 +261,6 @@ def read_scheduler
       # very long starting time of the yast module because the Storage module
       # (which is imported by the Bootloader (imported by the SystemSettings
       # module)) has a Read() function call in its constructor.
-      Yast.import "Bootloader"
```
Comment (jreidinger) [inline src/modules/SystemSettings.rb]: comment above no longer make sense

**Comment by jreidinger** [comment]
LGTM

**Pull request merged**